pkg/controllers/trace_controller_test.go

```diff
@@ -0,0 +1,208 @@
+// Copyright 2021 The Inspektor Gadget authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package controllers
+
+import (
+	"context"
+	"fmt"
+	"sort"
+	"sync"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+	. "github.com/onsi/gomega/types"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	gadgetv1alpha1 "github.com/kinvolk/inspektor-gadget/pkg/api/v1alpha1"
+	"github.com/kinvolk/inspektor-gadget/pkg/gadgets"
+)
+
+// FakeGadget is a fake implementation of the Gadget interface for tests. It records
+// the calls to its methods for assertions in the unit tests
+type FakeGadget struct {
+	mu    sync.Mutex
+	calls map[string]struct{}
+}
+
+func (g *FakeGadget) Delete(trace types.NamespacedName) error {
+	g.mu.Lock()
+	g.calls["delete/"+trace.String()] = struct{}{}
+	g.mu.Unlock()
+	return nil
+}
+
+func (g *FakeGadget) Operation(trace *gadgetv1alpha1.Trace, operation string, params map[string]string) error {
+	g.mu.Lock()
+	key := fmt.Sprintf("operation/%s/%s/%s/",
+		trace.ObjectMeta.Namespace,
+		trace.ObjectMeta.Name,
+		operation,
+	)
+	paramsKeys := make([]string, 0, len(params))
+	for k := range params {
+		paramsKeys = append(paramsKeys, k)
+	}
+	sort.Strings(paramsKeys)
+
+	for _, k := range paramsKeys {
+		key += fmt.Sprintf("%s/%s/", k, params[k])
+	}
+	g.calls[key] = struct{}{}
+	g.mu.Unlock()
+
+	trace.Status.OperationError = "FakeError"
+	trace.Status.State = "FakeState"
+	trace.Status.Output = "FakeOutput"
+	return nil
+}
+
+// methodHasBeenCalled is a helper function to check if a method has been
+// called on the gadget
+func methodHasBeenCalled(fakeGadget *FakeGadget, key string) func() string {
+	return func() string {
+		fakeGadget.mu.Lock()
+		defer fakeGadget.mu.Unlock()
+		_, ok := fakeGadget.calls[key]
+		if ok {
+			delete(fakeGadget.calls, key)
+			return key
+		} else {
+			return fmt.Sprint(fakeGadget.calls)
+		}
+	}
+}
```

Comment:
Couldn't it just be something like:

```
func (g *FakeGadget)methodHasBeenCalled(key string) bool {
		fakeGadget.mu.Lock()
		defer fakeGadget.mu.Unlock()
		_, ok := fakeGadget.calls[key]
               return ok
}
```

And `OperationMethodHasBeenCalled` be something like:

```
// OperationMethodHasBeenCalled returns a Gomega assertion checking if the
// method Operation() has been called
func OperationMethodHasBeenCalled(fakeGadget *FakeGadget, name, operation, params string) func() bool {
	key := fmt.Sprintf("operation/%s/%s/%s",
		name,
		operation,
		params,
	)
	return func() bool {
		return fakeGadget.methodHasBeenCalled(key)
	}
}
```
Am I missing something obvious?

Reply:
Thanks for the hint! Yes, that works.